File: src/stacklight.ts
import path from 'path';
import SerialPort = require('serialport');
import { promisify } from 'util';
const sleep = promisify(setTimeout);

export class StackLight {
  private port: any;
  private startByte: number;
  private endByte: number;
  private currentLightMode: number;
  private currentFlashFrequency: number;
  private currentBuzzerMode: number;

  constructor(port: string = 'COM3', baudRate: number = 9600) {
    this.startByte = 0xFF;
    this.endByte = 0xAA;
    this.currentLightMode = 0x04; // Default red
    this.currentFlashFrequency = 0x01; // No blinking by default
    this.currentBuzzerMode = 0x01; // Buzzer off by default
    
    try {
      this.port = new SerialPort.SerialPort({ path: port, baudRate: baudRate });
      this.port.on('error', (err:any) => {
        console.error('Serial port error:', err.message);
        this.port = null; // Disconnect if there’s an error
      });
      console.log(`Connected to ${port} at baud rate ${baudRate}`);
    } catch (error) {
      console.error("Failed to connect to Stacklight:", error);
      this.port = null; // Set port to null if connection fails
    }
  }
  

  private async sendCommand(): Promise<void> {
    // Prepare the command
    const command = Buffer.from([
      this.startByte,
      this.currentLightMode,
      this.currentBuzzerMode,
      this.currentFlashFrequency,
      this.endByte
    ]);

    // Send the command over the serial connection
    try {
      this.port.write(command);      
    } catch (error) {
      console.log("error with stacklight")
    }
    await sleep(500); // Wait to ensure the command is sent
  }

  public setLightSync(color: string): void {
    // Set the light color synchronously
    const colorModes: { [key: string]: number } = {
      off: 0x01,
      green: 0x02,
      blue: 0x03,
      red: 0x04,
      cyan: 0x05,
      yellow: 0x06,
      magenta: 0x07,
      white: 0x08
    };

    if (colorModes[color.toLowerCase()]) {
      this.currentLightMode = colorModes[color.toLowerCase()];
    } else {
      throw new Error(`Unknown color: ${color}`);
    }

    try {
    // Send the command synchronously
    this.port.write(Buffer.from([
      this.startByte,
      this.currentLightMode,
      this.currentBuzzerMode,
      this.currentFlashFrequency,
      this.endByte
    ]));      
    } catch (error) {
      
    }

  }

  public async setLight(color: string): Promise<void> {
    // Set the light color asynchronously
    this.setLightSync(color);
    await this.sendCommand();
  }

  public setFlashSync(flash: string): void {
    // Set the flash mode synchronously
    const flashModes: { [key: string]: number } = {
      none: 0x01,
      fast: 0x02,
      normal: 0x03,
      slow: 0x04
    };

    if (flashModes[flash]) {
      this.currentFlashFrequency = flashModes[flash];
    } else {
      throw new Error(`Unknown flash mode: ${flash}`);
    }
    try {
    // Send the command synchronously
      this.port.write(Buffer.from([
        this.startByte,
        this.currentLightMode,
        this.currentBuzzerMode,
        this.currentFlashFrequency,
        this.endByte
      ]));      
    } catch (error) {
      
    }


  }

  public async setFlash(flash: string): Promise<void> {
    // Set the flash mode asynchronously
    this.setFlashSync(flash);
    await this.sendCommand();
  }

  public toggleBuzzerSync(state: boolean): void {
    // Toggle the buzzer synchronously
    this.currentBuzzerMode = state ? 0x02 : 0x01;
    try {
    // Send the command synchronously
      this.port.write(Buffer.from([
        this.startByte,
        this.currentLightMode,
        this.currentBuzzerMode,
        this.currentFlashFrequency,
        this.endByte
      ]));      
    } catch (error) {
      
    }

  }

  public async toggleBuzzer(state: boolean): Promise<void> {
    // Toggle the buzzer asynchronously
    this.toggleBuzzerSync(state);
    await this.sendCommand();
  }

  public async cycleColors(delay: number = 1000): Promise<void> {
    // Cycle through colors with a delay in between
    const colorModes = [0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08]; // Green, Blue, Red, Cyan, Yellow, Magenta, White
    for (const mode of colorModes) {
      this.currentLightMode = mode;
      await this.sendCommand();
      await sleep(delay);
    }
  }

  public close(): void {
    // Close the serial connection
    this.setLightSync('off');
    this.toggleBuzzerSync(false);
    try {
      this.port.close();
    } catch (error) {
      
    }
  }
}

/*
// Usage of the StackLight class
(async () => {
  const stackLight = new StackLight('COM3', 9600);

  try {
    // Set the color to green synchronously
    stackLight.setLightSync('green');

    // Set the flash mode to fast asynchronously
    await stackLight.setFlash('fast');

    // Enable the buzzer synchronously
    stackLight.toggleBuzzerSync(true);
    stackLight.toggleBuzzerSync(false);

    // Cycle through colors with a 2-second delay
    await stackLight.cycleColors(2000);
  } finally {
    // Close the serial connection
    stackLight.close();
  }

  console.log('Actions completed.');
})();*/

File: stacklight.py
import serial
import time

class StackLight:
    def __init__(self, port='COM3', baudrate=9600):
        """Initialisiert die serielle Verbindung und setzt Standardwerte für den Port und die Baudrate."""
        self.port = port
        self.baudrate = baudrate
        self.ser = serial.Serial(self.port, self.baudrate)
        self.start_byte = 0xFF
        self.end_byte = 0xAA
        self.current_light_mode = 0x04  # Standardmäßig rot
        self.current_flash_frequency = 0x01  # Kein Blinken standardmäßig
        self.current_buzzer_mode = 0x01  # Buzzer aus standardmäßig
    
    def __del__(self):
        self.close()

    
    def send_command(self):
        """Sendet den aktuellen Zustand des Stacklight (Farbe, Flash, Buzzer)."""
        command = [self.start_byte, self.current_light_mode, self.current_buzzer_mode, self.current_flash_frequency, self.end_byte]
        self.ser.write(bytearray(command))
        time.sleep(0.5)
    
    def set_light(self, color):
        """Setzt die Farbe des Lichts."""
        color_modes = {
            "off": 0x01,
            "green": 0x02,
            "blue": 0x03,
            "red": 0x04,
            "cyan": 0x05,
            "yellow": 0x06,
            "magenta": 0x07,
            "white": 0x08
        }
        
        if color.lower() in color_modes:
            self.current_light_mode = color_modes[color.lower()]
        else:
            raise ValueError(f"Unbekannte Farbe: {color}")
        
        self.send_command()
    
    def set_flash(self, flash):
        """Setzt den Flash-Modus."""
        flash_modes = {
            "none": 0x01,
            "fast": 0x02,
            "normal": 0x03,
            "slow": 0x04
        }
        
        if flash in flash_modes:
            self.current_flash_frequency = flash_modes[flash]
        else:
            raise ValueError(f"Unbekannter Flash-Modus: {flash}")
        
        self.send_command()
    
    def toggle_buzzer(self, state):
        """Schaltet den Buzzer ein oder aus."""
        self.current_buzzer_mode = 0x02 if state else 0x01
        self.send_command()
        time.sleep(0.5)
        
    def trigger_buzzer(self):
        stack_light.toggle_buzzer(True)
        stack_light.toggle_buzzer(False)


    
    def cycle_colors(self):
        """Wechselt die Farben des Lichts durch."""
        color_modes = [0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08]  # Grün, Blau, Rot, Cyan, Gelb, Magenta, Weiß
        for mode in color_modes:
            time.sleep(2)
            self.current_light_mode = mode
            self.send_command()
    
    def close(self):
        """Schließt die serielle Verbindung."""
        self.set_light("off")
        self.set_flash("none")
        self.toggle_buzzer(False)
        self.ser.close()

# Verwendung der StackLight-Klasse
if __name__ == "__main__":
    stack_light = StackLight(port='COM3', baudrate=9600)
    
    try:
        # Farbe setzen
        stack_light.set_light("green")
        
        # Flash-Modus einstellen
        stack_light.set_flash("fast")
        
        # Buzzer aktivieren
        stack_light.trigger_buzzer()

        
        # Farbwechsel starten
        stack_light.cycle_colors()
        
    finally:
        # Serielle Verbindung schließen
        print("finaly")
        #stack_light.close()

    print("Aktionen abgeschlossen.")